frontend/public/components/catalog-items.jsx

```diff
@@ -5,37 +5,126 @@ import {Link} from 'react-router-dom';
 import {CatalogTileView} from 'patternfly-react-extensions/dist/esm/components/CatalogTileView';
 import {CatalogTile} from 'patternfly-react-extensions/dist/esm/components/CatalogTile';
 import {VerticalTabs} from 'patternfly-react-extensions/dist/esm/components/VerticalTabs';
+import {FilterSidePanel} from 'patternfly-react-extensions/dist/esm/components/FilterSidePanel';
+import {EmptyState} from 'patternfly-react/dist/esm/components/EmptyState';
+import FormControl from 'patternfly-react/dist/esm/components/Form/FormControl';
 
 import {normalizeIconClass} from './catalog-item-icon';
-import {categorizeItems} from './utils/categorize-catalog-items';
+import {categorizeItems, reCategorizeItems} from './utils/categorize-catalog-items';
 
 export class CatalogTileViewPage extends React.Component {
   constructor(props) {
     super(props);
 
-    this.state = {
-      showAllItemsForCategory: null,
-      activeTabs: ['all'], // array of tabs [main category, sub-category]
+    const allCategories = categorizeItems(props.items);
+    this.initialCategories = _.cloneDeep(allCategories);
+    const activeTabs = ['all']; // array of tabs [main category, sub-category]
+    const filters = {
+      byName: {
+        value: '',
+        active: false,
+      },
+      byType: {
+        clusterServiceClass: {
+          label: 'Service Class',
+          value: 'ClusterServiceClass',
+          active: false,
+        },
+        imageStream: {
+          label: 'Source-to-Image',
+          value: 'ImageStream',
+          active: false,
+        },
+      }
+    };
+
+    let filterCounts = {
+      byType: {
+        clusterServiceClasses: 0,
+        imageStreams: 0,
+      },
     };
+
+    this.state = this.getCategoryState(activeTabs, allCategories);
+    filterCounts = this.getFilterCounts(activeTabs, filters, filterCounts, allCategories);
+
+    _.assign(this.state, {
+      showAllItemsForCategory: null,
+      activeTabs,
+      filters,
+      filterCounts,
+    });
   }
 
-  static getDerivedStateFromProps(props) {
-    const allCategories = categorizeItems(props.items);
+  componentDidUpdate(prevProps, prevState) {
+    const { filters, filterCounts, activeTabs, allCategories } = this.state;
+    const { items } = this.props;
+
+    if (items !== prevProps.items) {
+      const newCategories = categorizeItems(items);
+      this.initialCategories = _.cloneDeep(newCategories);
+      this.setState(this.getCategoryState(activeTabs, newCategories));
+    }
+
+    if (filters !== prevState.filters) {
+      const filteredItems = this.filterItems();
+      const newCategories = _.cloneDeep(reCategorizeItems(filteredItems, this.initialCategories));
```

Comment:
I worry about performance here if the catalog ever grows to hundreds / thousands of items. We might need to revisit this, but OK for now.

Reply:
Shouldn't `recategorizeItems` clone the categories *before* modifying it?